website/public/js/services/userServices.js

```diff
@@ -0,0 +1,416 @@
+      }
+
+      user._wrapped = false;
+
+      function sync() {
+        $http({
+          method: "GET",
+          url: 'api/v3/user/',
+        })
+        .then(function (response) {
+          Notification.text(response.data.message);
+
+          _.extend(user, response.data.data);
+
+          if (!user._wrapped) {
+            // This wraps user with `ops`, which are functions shared both on client and mobile. When performed on client,
+            // they update the user in the browser and then send the request to the server, where the same operation is
+            // replicated. We need to wrap each op to provide a callback to send that operation
+            $window.habitrpgShared.wrap(user);
+            _.each(user.ops, function(op,k){
+              user.ops[k] = function(req,cb){
+                if (cb) return op(req,cb);
+                op(req,function(err,response) {
+                  for(var updatedItem in req.body) {
+                    var itemUpdateResponse = userNotifications[updatedItem];
+                    if(itemUpdateResponse) Notification.text(itemUpdateResponse);
+                  }
+                  if (err) {
+                    var message = err.code ? err.message : err;
+                    if (MOBILE_APP) Notification.push({type:'text',text:message});
+                    else Notification.text(message);
+                    // In the case of 200s, they're friendly alert messages like "Your pet has hatched!" - still send the op
+                    if ((err.code && err.code >= 400) || !err.code) return;
+                  }
+                  userServices.log({op:k, params: req.params, query:req.query, body:req.body});
+                });
+              }
+            });
+          }
+
+          save();
+          $rootScope.$emit('userSynced');
+        })
+      }
+      sync();
+
+      var save = function () {
+        localStorage.setItem(STORAGE_USER_ID, JSON.stringify(user));
+        localStorage.setItem(STORAGE_SETTINGS_ID, JSON.stringify(settings));
+      };
+
+      function callOpsFunctionAndRequest (opName, endPoint, method, paramString, opData) {
+        if (!opData) opData = {};
+        $window.habitrpgShared.ops[opName](user, opData);
+
+        var url = 'api/v3/user/' + endPoint;
+        if (paramString) {
+          url += '/' + paramString
+        }
+
+        var body = {};
+        if (opData.body) body = opData.body;
+
+        var queryString = '';
+        if (opData.query) queryString = '?' + $.param(opData.query)
+
+        $http({
+          method: method,
+          url: url + queryString,
+          body: body,
+        })
+        .then(function (response) {
+          Notification.text(response.data.message);
+          save();
+        })
+      }
+
+      function setUser(updates) {
+        for (var key in updates) {
+          user[key] = updates[key];
+        }
+
+        sync();
```

Comment:
why sync() here? the changes aren't applied directly to the client data? it should not be necessary to get the user from the sever